File: claw-os-sdk/go/objects.go
package clawossdk

import (
	"net/url"
	"strings"
	"unicode"
	"unicode/utf8"
)

const maxObjectRefURIBytes = 4096

// ObjectRefError reports invalid object-reference components or canonical spelling.
type ObjectRefError struct{ Message string }

func (e *ObjectRefError) Error() string { return "invalid App object reference: " + e.Message }

func validateObjectReference(value ObjectRef) error {
	if !objectComponent(value.AppId, 128) {
		return &ObjectRefError{"app_id must be a lowercase ASCII component of at most 128 bytes"}
	}
	if !objectComponent(value.ObjectType, 64) {
		return &ObjectRefError{"object_type must be a lowercase ASCII component of at most 64 bytes"}
	}
	if !objectOpaque(value.ObjectId, 1024) {
		return &ObjectRefError{"object_id must contain 1..=1024 UTF-8 bytes without control characters"}
	}
	if value.Revision != nil && !objectOpaque(*value.Revision, 128) {
		return &ObjectRefError{"revision must contain 1..=128 UTF-8 bytes without control characters"}
	}
	return nil
}

// FormatReference formats an opaque identifier without resolving it or granting access.
func FormatReference(value ObjectRef) (string, error) {
	if err := validateObjectReference(value); err != nil {
		return "", err
	}
	uri := "app://" + value.AppId + "/" + value.ObjectType + "?id=" + encodeObjectQuery(value.ObjectId)
	if value.Revision != nil {
		uri += "&revision=" + encodeObjectQuery(*value.Revision)
	}
	if len(uri) > maxObjectRefURIBytes {
		return "", &ObjectRefError{"URI exceeds 4096 UTF-8 bytes"}
	}
	return uri, nil
}

// ParseReference accepts only canonical app:// references and performs no I/O.
func ParseReference(value string) (ObjectRef, error) {
	if len(value) > maxObjectRefURIBytes {
		return ObjectRef{}, &ObjectRefError{"URI exceeds 4096 UTF-8 bytes"}
	}
	if !utf8.ValidString(value) {
		return ObjectRef{}, &ObjectRefError{"URI must be valid UTF-8"}
	}
	rest, found := strings.CutPrefix(value, "app://")
	if !found {
		return ObjectRef{}, &ObjectRefError{"URI must use the exact app:// scheme"}
	}
	address, query, found := strings.Cut(rest, "?")
	if !found {
		return ObjectRef{}, &ObjectRefError{"URI requires an id query parameter"}
	}
	appID, objectType, found := strings.Cut(address, "/")
	if !found {
		return ObjectRef{}, &ObjectRefError{"URI requires one App and one object type"}
	}
	reference := ObjectRef{AppId: appID, ObjectType: objectType}
	idPresent := false
	for _, parameter := range strings.Split(query, "&") {
		key, encoded, found := strings.Cut(parameter, "=")
		if !found {
			return ObjectRef{}, &ObjectRefError{"invalid query parameter"}
		}
		switch {
		case key == "id" && !idPresent:
			decoded, err := decodeObjectQuery(encoded)
			if err != nil {
				return ObjectRef{}, err
			}
			reference.ObjectId = decoded
			idPresent = true
		case key == "revision" && reference.Revision == nil:
			decoded, err := decodeObjectQuery(encoded)
			if err != nil {
				return ObjectRef{}, err
			}
			reference.Revision = &decoded
		default:
			return ObjectRef{}, &ObjectRefError{"unknown or repeated query parameter"}
		}
	}
	if !idPresent {
		return ObjectRef{}, &ObjectRefError{"URI requires an id query parameter"}
	}
	canonical, err := FormatReference(reference)
	if err != nil {
		return ObjectRef{}, err
	}
	if canonical != value {
		return ObjectRef{}, &ObjectRefError{"URI is not canonically spelled"}
	}
	return reference, nil
}

func objectComponent(value string, limit int) bool {
	if len(value) == 0 || len(value) > limit || value[0] < 'a' || value[0] > 'z' {
		return false
	}
	for index := 0; index < len(value); index++ {
		c := value[index]
		if !(c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '_' || c == '-') {
			return false
		}
	}
	return true
}

func objectOpaque(value string, limit int) bool {
	if len(value) == 0 || len(value) > limit || !utf8.ValidString(value) {
		return false
	}
	for _, character := range value {
		if unicode.IsControl(character) {
			return false
		}
	}
	return true
}

func encodeObjectQuery(value string) string {
	return strings.ReplaceAll(url.QueryEscape(value), "+", "%20")
}

func decodeObjectQuery(value string) (string, error) {
	for index := 0; index < len(value); index++ {
		c := value[index]
		if c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || strings.ContainsRune("-._~", rune(c)) {
			continue
		}
		if c != '%' || index+2 >= len(value) || !objectHex(value[index+1]) || !objectHex(value[index+2]) {
			return "", &ObjectRefError{"query values must use RFC3986 percent encoding"}
		}
		index += 2
	}
	decoded, err := url.QueryUnescape(value)
	if err != nil || !utf8.ValidString(decoded) {
		return "", &ObjectRefError{"query value is not valid UTF-8"}
	}
	return decoded, nil
}

func objectHex(value byte) bool {
	return value >= '0' && value <= '9' || value >= 'A' && value <= 'F' || value >= 'a' && value <= 'f'
}

File: claw-os-sdk/python/src/claw_os_sdk/objects.py
"""Pure, canonical references to App-owned objects; never discovery or dispatch."""

from __future__ import annotations

import re
import unicodedata
from urllib.parse import quote, unquote_to_bytes

from .generated import ObjectRef, WireDecodeError, validate_object_ref

__all__ = ["ObjectRef", "ObjectRefError", "format_reference", "parse_reference"]

_COMPONENT = re.compile(r"[a-z][a-z0-9_-]*")
_QUERY_VALUE = re.compile(r"(?:[A-Za-z0-9._~-]|%[A-Fa-f0-9]{2})*")
_MAX_URI_BYTES = 4096


class ObjectRefError(ValueError):
    """An invalid object-reference shape, component, or canonical URI."""


def _utf8_length(value: str) -> int:
    try:
        return len(value.encode("utf-8", errors="strict"))
    except UnicodeEncodeError as error:
        raise ObjectRefError("object reference text must be valid UTF-8") from error


def _validate(value: ObjectRef) -> None:
    try:
        validate_object_ref(value)
    except WireDecodeError as error:
        raise ObjectRefError(str(error)) from error
    for field, text, limit in (
        ("app_id", value["app_id"], 128),
        ("object_type", value["object_type"], 64),
    ):
        if len(text) > limit or _COMPONENT.fullmatch(text) is None:
            raise ObjectRefError(
                f"{field} must be a lowercase ASCII component of at most {limit} bytes"
            )
    for field, text, limit in (
        ("object_id", value["object_id"], 1024),
        ("revision", value.get("revision"), 128),
    ):
        if text is None:
            continue
        if (
            not text
            or len(text) > limit
            or _utf8_length(text) > limit
            or any(unicodedata.category(character) == "Cc" for character in text)
        ):
            raise ObjectRefError(
                f"{field} must contain 1..={limit} UTF-8 bytes without control characters"
            )


def format_reference(value: ObjectRef) -> str:
    """Format an identifier without trimming it, resolving data, or granting access."""
    _validate(value)
    uri = (
        f"app://{value['app_id']}/{value['object_type']}?id="
        + quote(value["object_id"], safe="-._~", encoding="utf-8", errors="strict")
    )
    if "revision" in value:
        uri += "&revision=" + quote(
            value["revision"], safe="-._~", encoding="utf-8", errors="strict"
        )
    if len(uri) > _MAX_URI_BYTES:
        raise ObjectRefError("URI exceeds 4096 UTF-8 bytes")
    return uri


def _decode(value: str) -> str:
    if _QUERY_VALUE.fullmatch(value) is None:
        raise ObjectRefError("query values must use RFC3986 percent encoding")
    try:
        return unquote_to_bytes(value).decode("utf-8", errors="strict")
    except UnicodeDecodeError as error:
        raise ObjectRefError("query value is not valid UTF-8") from error


def parse_reference(value: str) -> ObjectRef:
    """Parse only canonical app:// references, without any I/O."""
    if not isinstance(value, str):
        raise ObjectRefError("URI must be a string")
    if len(value) > _MAX_URI_BYTES or _utf8_length(value) > _MAX_URI_BYTES:
        raise ObjectRefError("URI exceeds 4096 UTF-8 bytes")
    if not value.startswith("app://"):
        raise ObjectRefError("URI must use the exact app:// scheme")
    address, separator, query = value[6:].partition("?")
    if not separator:
        raise ObjectRefError("URI requires an id query parameter")
    app_id, separator, object_type = address.partition("/")
    if not separator:
        raise ObjectRefError("URI requires one App and one object type")
    parameters: dict[str, str] = {}
    for parameter in query.split("&"):
        key, separator, encoded = parameter.partition("=")
        if not separator or key not in ("id", "revision") or key in parameters:
            raise ObjectRefError("unknown, repeated, or invalid query parameter")
        parameters[key] = _decode(encoded)
    if "id" not in parameters:
        raise ObjectRefError("URI requires an id query parameter")
    reference: ObjectRef = {
        "app_id": app_id,
        "object_type": object_type,
        "object_id": parameters["id"],
    }
    if "revision" in parameters:
        reference["revision"] = parameters["revision"]
    if format_reference(reference) != value:
        raise ObjectRefError("URI is not canonically spelled")
    return reference
